File: fe/src/domains/interaction/infrastructure/sessionEvent.ts
/**
 * 업무 이벤트 봉투. 백엔드 `SessionEvent` 와 같은 형태다(티켓 63 계약).
 *
 * 세션 주제 하나(`/topic/sessions/{sessionId}`)로 모든 종류가 내려오고 `type` 으로 갈라 처리한다.
 * 손들기·반응(64), 화면 공유(65), 강제 음소거(66)가 같은 봉투에 종류를 추가한다.
 */

/** 지금은 채팅뿐이다. 64~66 이 값을 추가한다. */
export type SessionEventType = "CHAT_MESSAGE";

export interface SessionEventSender {
  /**
   * LiveKit participant identity 와 **같은 값**(`p-{participantId}`). 참가자 목록과 이 이벤트를
   * 이어 붙이는 키다. 이메일 등 개인 식별 정보는 담기지 않는다.
   */
  readonly identity: string;
  readonly displayName: string;
  readonly role: "INSTRUCTOR" | "STUDENT";
}

export interface SessionEventEnvelope {
  /** 서버가 부여한 식별자. 중복 제거 기준이고 값 자체가 발생 순서를 담는다. */
  readonly eventId: string;
  /** 내가 보낸 것이면 내가 만든 값이 그대로 돌아온다. 다른 사람의 이벤트에도 값이 실려 있다. */
  readonly clientEventId: string | null;
  readonly type: SessionEventType;
  readonly sender: SessionEventSender;
  /** 수업 시작 기준 발생 시각(ms). 리포트 타임라인과 같은 축이다. */
  readonly occurredOffsetMs: number;
  readonly deliveredAt: string;
  readonly payload: Readonly<Record<string, unknown>>;
}

/** 전송이 거절됐다는 통지. 보낸 사람에게만 온다(`/user/queue/errors`). */
export interface SessionEventRejection {
  readonly clientEventId: string;
  readonly reason: string;
}

const isNonBlankString = (value: unknown): value is string =>
  typeof value === "string" && value.trim().length > 0;

const parseSender = (value: unknown): SessionEventSender | null => {
  if (typeof value !== "object" || value === null) return null;
  const sender = value as Record<string, unknown>;
  if (!isNonBlankString(sender.identity) || !isNonBlankString(sender.displayName)) return null;
  return {
    identity: sender.identity,
    displayName: sender.displayName,
    role: sender.role === "INSTRUCTOR" ? "INSTRUCTOR" : "STUDENT",
  };
};

/**
 * 봉투를 읽는다. 형태가 어긋나면 null 을 준다 — 알 수 없는 프레임 하나로 수업 화면이 죽지 않게 한다.
 *
 * 모르는 `type` 도 null 이다. 64~66 이 배포되는 사이 이 화면이 먼저 떠 있을 수 있다.
 */
export function parseSessionEvent(body: string): SessionEventEnvelope | null {
  let raw: unknown;
  try {
    raw = JSON.parse(body);
  } catch {
    return null;
  }
  if (typeof raw !== "object" || raw === null) return null;

  const event = raw as Record<string, unknown>;
  const sender = parseSender(event.sender);
  if (
    !isNonBlankString(event.eventId) ||
    event.type !== "CHAT_MESSAGE" ||
    sender === null ||
    typeof event.occurredOffsetMs !== "number"
  ) {
    return null;
  }

  return {
    eventId: event.eventId,
    clientEventId: isNonBlankString(event.clientEventId) ? event.clientEventId : null,
    type: event.type,
    sender,
    occurredOffsetMs: event.occurredOffsetMs,
    deliveredAt: typeof event.deliveredAt === "string" ? event.deliveredAt : "",
    payload:
      typeof event.payload === "object" && event.payload !== null
        ? (event.payload as Record<string, unknown>)
        : {},
  };
}

export function parseSessionEventRejection(body: string): SessionEventRejection | null {
  let raw: unknown;
  try {
    raw = JSON.parse(body);
  } catch {
    return null;
  }
  if (typeof raw !== "object" || raw === null) return null;

  const rejection = raw as Record<string, unknown>;
  if (!isNonBlankString(rejection.clientEventId)) return null;
  return {
    clientEventId: rejection.clientEventId,
    reason: isNonBlankString(rejection.reason) ? rejection.reason : "SEND_FAILED",
  };
}

/** 채팅 본문. 계약상 `payload.content` 하나다. */
export function chatContentOf(event: SessionEventEnvelope): string | null {
  return isNonBlankString(event.payload.content) ? event.payload.content : null;
}

File: backend/src/main/java/com/a105/zani/session/application/port/SessionEvent.java
package com.a105.zani.session.application.port;

import java.time.Instant;
import java.util.Map;

/**
 * 업무 이벤트 공통 봉투. 세션 주제 하나로 모든 종류가 나가고, 클라이언트는 {@code type} 으로 갈라 처리한다.
 *
 * <p>손들기·반응(64), 화면 공유(65), 강제 음소거(66)가 같은 봉투를 재사용한다. 종류마다 봉투를 따로 만들면 클라이언트가 중복 제거·순서 처리를 종류별로 반복해야 한다.
 *
 * @param eventId 서버가 부여한 식별자(TSID 문자열). 중복 제거의 기준이고, 값 자체가 발생 순서를 담는다. 자바스크립트의 정수 정밀도를 넘지 않도록 문자열로 보낸다.
 * @param clientEventId 클라이언트가 만들어 보낸 값을 그대로 되돌려 준다. 보낸 쪽이 낙관적으로 그려 둔 항목을 확정된 이벤트와 맞추는 데 쓴다. 서버는 저장하지 않는다 —
 *     {@code chat_messages} 에 담을 컬럼이 없고, 재시도 멱등은 Redis 로 처리한다.
 * @param occurredOffsetMs 수업 시작으로부터 흐른 밀리초. 리포트 타임라인과 같은 축이다. 서버가 계산한다.
 * @param deliveredAt 서버가 내보낸 시각(UTC). 화면에 시각을 표시할 때 쓴다.
 * @param payload 종류별 추가 데이터. 채팅은 {@code content} 하나다.
 */
public record SessionEvent(
        String eventId,
        String clientEventId,
        SessionEventType type,
        SessionEventSender sender,
        long occurredOffsetMs,
        Instant deliveredAt,
        Map<String, Object> payload) {

    public SessionEvent {
        payload = payload == null ? Map.of() : Map.copyOf(payload);
    }
}
